XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+an asynchronous event they will continue to fire for all nested asynchronous
+events. Even after the instance has been disabled.
+
+Callbacks are not implicitly enabled after an instance is created. Reason for
+this is to not make any assumptions about the user's use case. Since
+constructing the `asyncHook` during startup but not using it until later is a
+perfectly reasonable use case. This API is meant to err on the side of
+requiring explicit instructions from the user.
+
+#### `asyncHook.disable()`
+
+Disable the callbacks for a given `AsyncHook` instance. Doing this will prevent
+the `init()`, etc., calls from firing for any new roots of asynchronous call
+stacks, but will not prevent existing asynchronous call stacks that have
+already been captured by the `AsyncHook` instance from continuing to fire.
+
+While not part of the immediate development plan, it should be possible in the
+future to allow selective tracking of asynchronous call stacks. The following
+example demonstrates this:
+
+```js
+const async_wrap = require('async_wrap');
+const net = require('net');
+
+// Pretend init, pre, post are all defined
+const asyncHook = async_wrap.createHook({ init, pre, post });
+asyncHook.enable();
+
+net.createServer((c) => {
+  // Only want to follow connections that match IP range.
+  if (ipRangeRegExp.test(c.address().address))
+    asyncHook.disable();
+}).listen(PORT);
+```
+
+At which point no further calls to the hooks on that instance will be made for
+that asynchronous branch.
+
+#### `asyncHook.scope()`
+
+Enable capture of asynchronous events until the current synchronous code
+execution has completed. This is basically a small amount of sugar to prevent
+accidentally forgetting to `disable()` a set of hooks in a function with
+multiple returns. Also in case an error is thrown and caught by a domain or
+`uncaughtException`.
+
+### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
+four areas. On instantiation, before/after the callback is called and when the
+instance is destructed. For cases where resources are reused, instantiation and
+destructor calls are emulated.
+
+#### `init(id, type, parentId)`
+
+* `id` {Number}
+* `type` {String}
+* `parentId` {Number}
+
+Called when a class is constructed that has the possibility to trigger an
+asynchronous event. This does mean the instance _will_ trigger a
+`pre()`/`post()` event. Only that the possibility exists.
+
+The `this` of each call is that of the object being constructed. While core
+attempts to make accessing properties of these handles safe at construction, it
+is not guaranteed (especially if third-party modules are involved) that
+everything can be safely accessed. So it is not recommended to blindly
+iterate over the object's properties. Despite this, it is important to give
+access of the handle to the user so they can utilize tracking methods of their
+choosing. For example:
```

Comment:
The argument against `this` is primarily one of convention — anecdotally, many JS APIs have been moving away from "load-bearing `this`" over the last few years. When I was poking at the asyncwrap API (pre-docs), I was surprised that `this` carried pertinent info — it was not obvious to me  that the hook was running as an extension of the handle.

It's admittedly a nit — but if it comes at zero cost & makes the API more obvious, it might be worth addressing.